examples/svdiff/modeling_svdiff.py

```diff
@@ -0,0 +1,133 @@
+                    return updated_weight.unsqueeze(0)
+                else:
+                    return updated_weight.squeeze(0)
+            else:
+                raise ValueError(f"`weight_type`={self.weight_type} is invalid!")
+        else:
+            return updated_weight
+        
+    def forward(self, weight: torch.Tensor):
+        updated_weight = self.U @ torch.diag(F.relu(self.S + self.scale * self.delta)) @ self.Vh
+        updated_weight = self._reshape_weight_for_svd(updated_weight, weight, reverse=True)
+        return updated_weight
+
+
+def set_spectral_shifts(model: nn.Module, spectral_shifts_ckpt: str = None, **kwargs):
+    # key to module
+    svdiff_modules = {}
+    for name, module in model.named_modules():
+        svdiff_module = None
+        if isinstance(module, nn.Conv1d) or isinstance(module, nn.Conv2d):
+            svdiff_module = SVDiffModule(module.weight, weight_type="conv")
+        elif isinstance(module, nn.LayerNorm) or isinstance(module, nn.GroupNorm):
+            svdiff_module = SVDiffModule(module.weight, weight_type="1d")
+        elif isinstance(module, nn.Linear) or isinstance(module, nn.Embedding):
+            svdiff_module = SVDiffModule(module.weight)
+        if svdiff_module:
+            svdiff_modules[name+".delta"] = svdiff_module
+            # register parametrization 
+            nn.utils.parametrize.register_parametrization(module, "weight", svdiff_module)
+            
+    if spectral_shifts_ckpt:
+        param_device = "cpu"
+        torch_dtype = kwargs["torch_dtype"] if "torch_dtype" in kwargs else None
+        with safetensors.safe_open(spectral_shifts_ckpt, framework="pt", device="cpu") as f:
+            for key in f.keys():
+                accepts_dtype = "dtype" in set(inspect.signature(set_module_tensor_to_device).parameters.keys())
+                if accepts_dtype:
+                    set_module_tensor_to_device(svdiff_modules[key], "delta", param_device, value=f.get_tensor(key), dtype=torch_dtype)
+                else:
+                    set_module_tensor_to_device(svdiff_modules[key], "delta", param_device, value=f.get_tensor(key))
+        print(f"Resumed from {spectral_shifts_ckpt}")
+    return model, svdiff_modules
+
+
+def slerp(val, low, high):
+    """ taken from https://discuss.pytorch.org/t/help-regarding-slerp-function-for-generative-model-sampling/32475/4
+    """
+    low_norm = low/torch.norm(low, dim=1, keepdim=True)
+    high_norm = high/torch.norm(high, dim=1, keepdim=True)
+    omega = torch.acos((low_norm*high_norm).sum(1))
+    so = torch.sin(omega)
+    res = (torch.sin((1.0-val)*omega)/so).unsqueeze(1)*low + (torch.sin(val*omega)/so).unsqueeze(1) * high
+    return res
+
+
+def slerp_tensor(val, low, high):
+    shape = low.shape
+    res = slerp(val, low.flatten(1), high.flatten(1))
+    return res.reshape(shape)
+
+
+def ddim_invert(pipe, prompt=None, image=None, num_inference_steps=50, guidance_scale=1.0, **kwargs):
```

Comment:
Can't we just use this method? 

https://github.com/huggingface/diffusers/blob/262d539a8a8f505dc72958f7ea50915a4b56dfac/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py#L1085

Reply:
Ah, probably okay for this PR.